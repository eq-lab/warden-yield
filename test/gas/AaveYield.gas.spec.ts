import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';
import snapshotGasCost from '@uniswap/snapshot-gas-cost';
import { createAaveEthFork } from '../shared/fixtures';
import { ethers } from 'hardhat';
import { parseEther, TransactionReceipt } from 'ethers';
import {
  USER_WARDEN_ADDRESS,
  WardenChain,
  WardenContractAddress,
  encodeStakeAction,
  encodeUnstakeAction,
  setTokenBalance,
} from '../shared/utils';
import { IERC20Metadata__factory } from '../../typechain-types';
import { CommandId } from '../shared/warden-handler-fixtures';

describe('AaveYield gas', () => {
  it('user stake', async () => {
    const { aaveYield, weth9 } = await loadFixture(createAaveEthFork);
    const [_, user] = await ethers.getSigners();

    const input = parseEther('1');
    await setTokenBalance(await weth9.getAddress(), user.address, input);

    await weth9.connect(user).transfer(aaveYield.target, input);
    const stakeId = 1;
    const stakePayload = encodeStakeAction(stakeId);
    const tokenSymbol = await IERC20Metadata__factory.connect(await weth9.getAddress(), user).symbol();

    // stake
    const txReceipt = await (
      await aaveYield
        .connect(user)
        .executeWithToken(CommandId, WardenChain, WardenContractAddress, stakePayload, tokenSymbol, input)
    ).wait();

    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });

  it('user unstake', async () => {
    const { aaveYield, weth9 } = await loadFixture(createAaveEthFork);
    const [_, user] = await ethers.getSigners();

    const input = parseEther('1');
    await setTokenBalance(await weth9.getAddress(), user.address, input);

    //stake
    await weth9.connect(user).transfer(aaveYield.target, input);
    const stakeId = 1;
    const stakePayload = encodeStakeAction(stakeId);
    const tokenSymbol = await IERC20Metadata__factory.connect(await weth9.getAddress(), user).symbol();

    await aaveYield
      .connect(user)
      .executeWithToken(CommandId, WardenChain, WardenContractAddress, stakePayload, tokenSymbol, input);

    //unstake
    const unstakeId = 1;
    const lpAmount = parseEther('0.5');
    const unstakePayload = encodeUnstakeAction(unstakeId, lpAmount);

    const txReceipt = await (
      await aaveYield.connect(user).execute(CommandId, WardenChain, WardenContractAddress, unstakePayload)
    ).wait();

    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });
});
