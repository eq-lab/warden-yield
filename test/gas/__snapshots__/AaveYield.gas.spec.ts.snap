// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AaveYield gas user stake 1`] = `656944`;

exports[`AaveYield gas user unstake 1`] = `590178`;
